src/DefaultBuilder/src/Configuration.cs

```diff
@@ -4,126 +4,145 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
 using Microsoft.Extensions.Configuration;
-using Microsoft.Extensions.FileProviders;
 using Microsoft.Extensions.Primitives;
 
-// TODO: Microsft.Extensions.Configuration API Proposal
 namespace Microsoft.AspNetCore.Builder
 {
     /// <summary>
-    /// Configuration is mutable configuration object. It is both a configuration builder and an IConfigurationRoot. 
+    /// Configuration is mutable configuration object. It is both an <see cref="IConfigurationBuilder"/> and an <see cref="IConfigurationRoot"/>.
     /// As sources are added, it updates its current view of configuration. Once Build is called, configuration is frozen.
     /// </summary>
-    public sealed class Configuration : IConfigurationRoot, IConfigurationBuilder
+    public sealed class Configuration : IConfigurationRoot, IConfigurationBuilder, IDisposable
     {
-        private readonly ConfigurationBuilder _builder = new();
-        private IConfigurationRoot _configuration;
+        private readonly ConfigurationSources _sources;
+        private ConfigurationRoot _configurationRoot;
 
-        /// <summary>
-        /// Gets or sets a configuration value.
-        /// </summary>
-        /// <param name="key">The configuration key.</param>
-        /// <returns>The configuration value.</returns>
-        public string this[string key] { get => _configuration[key]; set => _configuration[key] = value; }
+        private ConfigurationReloadToken _changeToken = new();
+        private IDisposable? _changeTokenRegistration;
 
         /// <summary>
-        /// Gets a configuration sub-section with the specified key.
+        /// Creates an empty  mutable configuration object that is both an <see cref="IConfigurationBuilder"/> and an <see cref="IConfigurationRoot"/>.
         /// </summary>
-        /// <param name="key">The key of the configuration section.</param>
-        /// <returns>The <see cref="IConfigurationSection"/>.</returns>
-        /// <remarks>
-        ///     This method will never return <c>null</c>. If no matching sub-section is found with the specified key,
-        ///     an empty <see cref="IConfigurationSection"/> will be returned.
-        /// </remarks>
-        public IConfigurationSection GetSection(string key)
+        public Configuration()
         {
-            return _configuration.GetSection(key);
+            _sources = new ConfigurationSources(this);
+
+            // Make sure there's some default storage since there are no default providers.
+            this.AddInMemoryCollection();
```

Comment:
Why is this in here vs being in the caller? This can't go into the BCL like this right?

Reply:
The indexer doesn't really work otherwise after it's newly constructed. The following test would fail if you didn't call `.AddInMemoryCollection()` yourself which I think is a bad experience:

https://github.com/dotnet/aspnetcore/pull/32926/files#diff-6dcb162a1522a52cc83721420a5af0212db9a6426a75eb91793c49085c0a1221R84-R96

I got the idea from HostBuilder:

https://github.com/dotnet/runtime/blob/b54dbaf9adb2ec068098394b3f73e8e7cd0ec09c/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L141-L142

Reply:
I hear you but you didn't answer my question

Reply:
> Why is this in here vs being in the caller?

Because it's more usable this way.

> This can't go into the BCL like this right?

I don't see why not. I copied this from the runtime repo.